core/src/util/pbfParser.cpp

```diff
@@ -139,102 +246,111 @@ Feature PbfParser::getFeature(ParserContext& _ctx, protobuf::message _featureIn)
-        if (tagValue >= 0) {
-            properties.emplace_back(_ctx.keys[tagKey], _ctx.values[tagValue]);
-        }
+    if (!extractTags(_ctx, tagsMsg)) {
+        return;
     }
-    feature.props.setSorted(std::move(properties));
 
-    switch(feature.geometryType) {
-        case GeometryType::points:
-            feature.points.insert(feature.points.begin(),
-                                  _ctx.geometry.coordinates.begin(),
-                                  _ctx.geometry.coordinates.end());
-            break;
+    if (!_sink.matchFeature(feature)) {
+        return;
+    }
 
-        case GeometryType::lines:
-        {
-            auto pos = _ctx.geometry.coordinates.begin();
-            for (int length : _ctx.geometry.sizes) {
+    extractGeometry(_ctx, geomMsg);
+
+    switch(feature.geometryType) {
+        case GeometryType::points: {
+                feature.points.clear();
+                feature.points.insert(feature.points.begin(),
+                                      _ctx.coordinates.begin(),
+                                      _ctx.coordinates.end());
+                break;
+        }
+        case GeometryType::lines: {
+            int offset = 0;
+            feature.lines.clear();
+            for (int length : _ctx.numCoordinates) {
                 if (length == 0) { continue; }
+
                 Line line;
                 line.reserve(length);
-                line.insert(line.begin(), pos, pos + length);
-                pos += length;
+
+                line.insert(line.begin(),
+                            _ctx.coordinates.begin() + offset,
+                            _ctx.coordinates.begin() + offset + length);
+
+                offset += length;
                 feature.lines.emplace_back(std::move(line));
             }
             break;
         }
-        case GeometryType::polygons:
-        {
-            auto pos = _ctx.geometry.coordinates.begin();
-            auto rpos = _ctx.geometry.coordinates.rend();
-            for (int length : _ctx.geometry.sizes) {
+        case GeometryType::polygons: {
+            int offset = 0;
+            WindingOrder windingOrder;
+
+            feature.polygons.clear();
+
+            for (int length : _ctx.numCoordinates) {
                 if (length == 0) { continue; }
-                float area = signedArea(pos, pos + length);
-                if (area == 0) {
-                    pos += length;
-                    rpos -= length;
+                auto it = _ctx.coordinates.begin() + offset;
+
+                double area = signedArea(it, it + length);
+                if (area == 0.f) {
+                    offset += length;
                     continue;
                 }
-                int winding = area > 0 ? 1 : -1;
-                // Determine exterior winding from first polygon.
-                if (_ctx.winding == 0) {
-                    _ctx.winding = winding;
+                WindingOrder winding(area);
```

Comment:
So, the logic here related to winding direction is intentionally not generalized. You can't generally say whether a positive signed area is clockwise or counterclockwise because it depends on the coordinate system of the points in the ring (notably, whether y is up or down). It's really simpler to avoid talking in terms of "clockwise" and "counterclockwise" altogether, since all that matters to us is the direction of one ring relative to another and the English terms can (and do) lead to confusion. The `WindingOrder` type is really unrelated to the purpose of this PR so I suggest it be omitted.